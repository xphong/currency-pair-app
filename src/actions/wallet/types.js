export const REQUEST_TRADE = `wallet/REQUEST_TRADE`;
export const COMPLETE_TRADE = `wallet/COMPLETE_TRADE`;
export const TRADE_ERROR = `wallet/TRADE_ERROR`;
