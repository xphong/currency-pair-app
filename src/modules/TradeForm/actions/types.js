export const REQUEST_QUOTE = `quote/REQUEST_QUOTE`;
export const RETRIEVE_QUOTE = `quote/RETRIEVE_QUOTE`;
export const QUOTE_ERROR = `quote/QUOTE_ERROR`;
