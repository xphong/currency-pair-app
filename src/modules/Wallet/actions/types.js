export const REQUEST_TRADE = `wallet/REQUEST_TRADE`;
export const RECEIVE_TRADE = `wallet/RECEIVE_TRADE`;
export const TRADE_ERROR = `wallet/TRADE_ERROR`;
