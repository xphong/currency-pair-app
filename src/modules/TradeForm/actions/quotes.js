import axios from 'axios';

import * as types from './types';

export const requestQuote = () => {
  return {
    type: types.REQUEST_QUOTE
  };
}

export const receiveQuote = data => {
  return {
    type: types.RECEIVE_QUOTE,
    payload: {
      data
    }
  };
}

export const receiveError = error => {
  return {
    type: types.QUOTE_ERROR,
    payload: {
      error
    }
  };
}

export const fetchBitcoinQuote = () => {
  const url = 'https://api.bitfinex.com/v1/pubticker/btcusd';
  const opts = {
    url,
    timeout: 10000,
    method: 'get',
    responseType: 'json'
  };

  return async dispatch => {
    dispatch(requestQuote());

    try {
      const { data } = await axios(opts);

      dispatch(receiveQuote(Number(data.last_price)));
    } catch (e) {
      dispatch(receiveError(e));
    }
  };
}
