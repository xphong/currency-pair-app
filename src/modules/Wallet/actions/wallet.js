import * as types from './types';

import delay from 'utils/delay/delay';

export const requestTrade = () => {
  return {
    type: types.REQUEST_TRADE
  };
}

export const receiveTrade = balance => {
  return {
    type: types.RECEIVE_TRADE,
    payload: {
      balance
    }
  };
}

export const receiveError = error => {
  return {
    type: types.TRADE_ERROR,
    payload: {
      error
    }
  };
}

export const trade = (amount, quoteAmount) => {
  return async (dispatch, getState) => {
    const { wallet } = getState();
    const usdBalance = wallet.balance.USD;
    const btcBalance = wallet.balance.BTC;

    if (!amount || !quoteAmount) {
      dispatch(receiveError('Please enter a trade amount.'));
      return;
    }

    if (amount > usdBalance) {
      dispatch(receiveError('You have insufficient funds in your account.'));
      return;
    }

    dispatch(requestTrade());

    try {
      // Fake a delay. In a real-world application, trade would make a POST call
      await delay(1000);

      const balance = {
        USD: usdBalance - amount,
        BTC: btcBalance + quoteAmount
      };

      dispatch(receiveTrade(balance));
    } catch (e) {
      dispatch(receiveError(e.message));
    }
  };
}
