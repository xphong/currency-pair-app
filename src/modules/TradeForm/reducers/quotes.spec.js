import * as types from '../actions/types';
import quotesReducer from './quotes';

describe('Quotes Reducer', () => {
  it('should return initial state', () => {
    const reducer = quotesReducer(undefined, {});

    expect(reducer).toMatchSnapshot();
  });

  it('should handle REQUEST_QUOTE', () => {
    const action = {
      type: types.REQUEST_QUOTE
    };

    const reducer = quotesReducer(undefined, action);

    expect(reducer).toMatchSnapshot();
  });

  it('should handle QUOTE_ERROR', () => {
    const action = {
      type: types.QUOTE_ERROR,
      payload: {
        data: {}
      }
    };

    const reducer = quotesReducer(undefined, action);

    expect(reducer).toMatchSnapshot();
  });

  it('should handle RECEIVE_QUOTE', () => {
    const action = {
      type: types.RECEIVE_QUOTE,
      payload: {
        data: '8288.9'
      }
    };

    const reducer = quotesReducer(undefined, action);

    expect(reducer).toMatchSnapshot();
  });
});
