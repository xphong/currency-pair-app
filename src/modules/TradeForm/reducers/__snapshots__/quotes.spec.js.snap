// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quotes Reducer should handle QUOTE_ERROR 1`] = `
Object {
  "bitcoin": 0,
  "error": undefined,
  "isLoading": false,
}
`;

exports[`Quotes Reducer should handle RECEIVE_QUOTE 1`] = `
Object {
  "bitcoin": "8288.9",
  "error": false,
  "isLoading": false,
}
`;

exports[`Quotes Reducer should handle REQUEST_QUOTE 1`] = `
Object {
  "bitcoin": 0,
  "error": false,
  "isLoading": true,
}
`;

exports[`Quotes Reducer should return initial state 1`] = `
Object {
  "bitcoin": 0,
  "error": false,
  "isLoading": false,
}
`;
